### ISSUE TITLE
Flaky test_gluon_model_zoo_gpu.test_training @ Python3: MKLDNN-GPU

### ISSUE THREAD

**Issue by marcoabreu** [apache/incubator-mxnet#9820]
http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/incubator-mxnet/detail/PR-9809/4/pipeline
http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/incubator-mxnet/detail/master/389/pipeline
http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/incubator-mxnet/detail/master/391/pipeline
http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/incubator-mxnet/detail/master/387/pipeline

```
======================================================================

FAIL: test_gluon_model_zoo_gpu.test_training

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/usr/local/lib/python3.5/dist-packages/nose/case.py", line 198, in runTest

    self.test(*self.arg)

  File "/workspace/tests/python/gpu/../unittest/common.py", line 155, in test_new

    orig_test(*args, **kwargs)

  File "/workspace/tests/python/gpu/test_gluon_model_zoo_gpu.py", line 159, in test_training

    assert_almost_equal(cpu_out.asnumpy(), gpu_out.asnumpy(), rtol=1e-2, atol=1e-2)

  File "/workspace/python/mxnet/test_utils.py", line 493, in assert_almost_equal

    raise AssertionError(msg)

AssertionError: 

Items are not equal:

Error 84.491165 exceeds tolerance rtol=0.010000, atol=0.010000.  Location of maximum error:(9, 877), a=0.816808, b=-0.181214

 a: array([[ 0.46908194,  0.25131682,  0.27432024, ..., -0.3243659 ,

        -0.8637756 ,  0.57461524],

       [ 0.18945426,  0.32339886,  0.18884647, ...,  0.00044107,...

 b: array([[ 0.34904164,  0.3221189 ,  0.05189171, ..., -0.14858764,

        -0.9381798 ,  0.39666444],

       [ 0.3438487 ,  0.4455883 ,  0.32494336, ...,  0.15454161,...

-------------------- >> begin captured logging << --------------------

urllib3.connectionpool: DEBUG: Starting new HTTP connection (1): data.mxnet.io

urllib3.connectionpool: DEBUG: http://data.mxnet.io:80 "GET /data/val-5k-256.rec HTTP/1.1" 200 150874780

root: INFO: downloaded http://data.mxnet.io/data/val-5k-256.rec into data/val-5k-256.rec successfully

common: INFO: Setting test np/mx/python random seeds, use MXNET_TEST_SEED=1760574333 to reproduce.

--------------------- >> end captured logging << ---------------------

```

**Comment by zheng-da** [comment]
I think we should disable it for now.
I run the tests many times. It seems both CPU and GPU occasionally generate
incorrect results. I need more time to test the tests.

**Comment by marcoabreu** [comment]
Sounds good. Since we've now got the CI randomness PR merged into the master, it should be no problem to generate faulty tests.

**Comment by marcoabreu** [comment]
Ah wait. I just checked a few runs of test failures and it ONLY happens with MKLDNN. It looks to me like this is not a flaky test but could rather have a problem in the underlying implementation.

@zheng-da @mli can you please make sure this gets resolved? This test is one of the biggest failure reasons and there might be a valid test failure.

**Comment by pengzhao-intel** [comment]
@marcoabreu Thanks for the testing. 
We will take look this case tomorrow (China, Beijing time zone) to verify this issue w/ high priority.
And will fix the issue in case it's a real issue introduced by the new code.

**Comment by ptrendx** [comment]
At least #9845 is not MKLDNN specific - I've seen (deterministic) failures in that test before, see #8780.

**Comment by zheng-da** [comment]
@marcoabreu I was testing it yesterday. It seems the problem exists when MKLDNN is compiled into MXNet, so it should be a bug caused by the code inside USE_MKLDNN. I'll look into the problem this week and try to fix it as soon as possible.

I agree with @pengzhao-intel . This problem might be related to the other flaky failures.

**Comment by zheng-da** [comment]
I run more tests. The bug occurs when MKLDNN is compiled into MXNet and we use the threaded engine, which is the default executing engine. If we use the naive executing engine, the problem doesn't show up. It means that the bug isn't in the operator implementation. It should be in the executor engine or in the NDArray.

**Comment by marcoabreu** [comment]
Thank you for the elaboration!

@szha @piiswrong what do you think?

**Comment by zheng-da** [comment]
Most likely, what happens is that an operation needs to wait for some variables before proceeding to the next operator, but somehow the implementation doesn't wait for the right variables or misses some variables.

**Comment by zheng-da** [comment]
After one day of testing, I think I know where the problem is now. Basically, the current implementation sometimes needs to convert data format (from the MKLDNN format to the default format) inside an NDArray. In the threaded execution engine, while an NDArray is being converted in a thread, the array can also be read by another thread. In this case, the other thread can read wrong data.

I changed the code to avoid data layout conversion inside an NDArray. It seems to work fine now. I'll have more tests.

**Comment by marcoabreu** [comment]
Great news, thanks for diving deep on this problem!

**Comment by marcoabreu** [comment]
Fixed as of #9862
